File: linkedlist/83_RemoveDuplicatesfromSortedList/83_RemoveDuplicatesfromSortedList.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def deleteDuplicates(self, head: Optional[ListNode]) -> Optional[ListNode]:
        
        if head == None:
            return head
        
        my_dic = {}
        temp = head

        while temp != None:            
            if my_dic.get(temp.val, -1) == -1 :               
                my_dic[temp.val] = temp
            
            temp = temp.next

        result = None
        prev = None
        
        for val in my_dic.values(): #for i, val in my_dic.items():
            if result == None:
                prev = val
                result = prev
                continue

            prev.next = val
            prev = prev.next
            
        prev.next = None    

        return result

File: linkedlist/83_RemoveDuplicatesfromSortedList/83_RemoveDuplicatesfromSortedList.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 */
var deleteDuplicates = function(head) {
    if(head == null)
        return head;

    let prev = head;    
    let temp = head;       
    temp = temp.next;    

    while(temp != null){    
        
        if(prev.val == temp.val){
            let temp2 = temp;    
            temp = temp.next;
            delete temp2;           
        }
        else{ 
           prev.next = temp;
           prev = prev.next;
           temp = temp.next
       }
    }

    if(prev != null && prev.next != null)
        prev.next = null;
    
    return head;
    
};